script/update-makefiles.py

```diff
@@ -0,0 +1,200 @@
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+# This script updates different make/build files (CMakeLists.txt, BUILD.gn,
+# Andriod.mk, Andriod.bp, auto-make) in OpenThread repo based on the
+# current files present in `./src/core/` & `./include/openthread/`
+# folders. This script MUST be called from openthread root folder.
+
+import os
+
+#----------------------------------------------------------------------------------------------
+# Helper functions
+
+
+def get_file_list(path, extension):
+    """Get a sorted list of full file names (with path) in a given `path` folder having a given `extension`"""
+    return sorted([
+        "{}/{}".format(dir_path, file_name)[2:]
+        for dir_path, dir_names, file_names in os.walk(path)
+        for file_name in file_names
+        if file_name.endswith(extension)
+    ])
+
+
+def read_txt_file(file_name):
+    """Read the content of a text file with name `file_name` and return content as list of lines"""
+    with open(file_name, 'r') as file:
+        lines = file.readlines()
+    return lines
+
+
+def write_txt_file(file_name, lines):
+    """Write a text file with name `file_name` with the content given as a list of `lines`"""
+    with open(file_name, 'w') as file:
+        file.writelines(lines)
+
+
+def update_build_file(file_name, start_string, end_string, new_list, search_string=None):
+    """
+    Update the file given by `file_name` by replacing the list after the first occurrence of `start_string` up to
+    `end_string` with the `new_list`. If `search_string` is given, then the search for `start_string` only starts
+    after seeing the `search_string` line in the file.
+    """
+    STATE_SEARCH = 1
+    STATE_MATCH_START = 2
+    STATE_MATCH_END = 3
+    STATE_DONE = 4
+
+    lines = read_txt_file(file_name)
+    new_lines = []
+    state = STATE_SEARCH if search_string else STATE_MATCH_START
+
+    for line in lines:
+        if state == STATE_SEARCH:
+            new_lines.append(line)
+            if line.startswith(search_string):
+                state = STATE_MATCH_START
+        elif state == STATE_MATCH_START:
+            new_lines.append(line)
+            if line.startswith(start_string):
+                new_lines.extend(new_list)
+                state = STATE_MATCH_END
+        elif state == STATE_MATCH_END:
+            if line.startswith(end_string):
+                new_lines.append(line)
+                state = STATE_DONE
+        else:
+            new_lines.append(line)
+
+    write_txt_file(file_name, new_lines)
```

Comment:
```suggestion
    if (state != STATE_DONE):
        raise RuntimeError('failed to update build file: %s' % file_name)

    if new_lines != lines: write_txt_file(file_name, new_lines)
```
Make sure we really updated the file?

Reply:
Good suggestions. Thanks.